### ISSUE TITLE
User and Role extensions can't be resolved

### ISSUE THREAD

**Issue by nurdyguy** [grandchamp/Identity.Dapper#21]
For my project I created extension models for each of the Identity.Dapper.Entities classes so that I can customize the various models.  However, when I then extended the stores and repos I run into problems because the stores and repos are based on the` DapperIdentityUser<int>` and `DapperIdentityRole<int>` classes, not the `DapperIdentityUser<TKey, TUserClaim, TUserRole, TUserLogin>` and `DapperIdentityRole<TKey, TUserRole, TRoleClaim>` classes.  

Ex:

    public class User : DapperIdentityUser<int, UserClaim, UserRole, UserLogin>
    {
        
    }

    public class UserRepository : UserRepository<User, int, UserRole, RoleClaim>, IUserRepository
    {
        private readonly IConnectionProvider _connectionProvider;
        private readonly ILogger<UserRepository> _log;
        private readonly SqlConfiguration _sqlConfiguration;
        private readonly IRoleRepository _roleRepository;
        //private readonly IRoleRepository<DapperIdentityRole<int>, int, UserRole, RoleClaim> _roleRepository;

        public UserRepository(IConnectionProvider connProv, 
                              ILogger<UserRepository> log, 
                              SqlConfiguration sqlConf,
                              //IRoleRepository<DapperIdentityRole<int>, int, UserRole, RoleClaim> roleRepo
                              IRoleRepository roleRepo
                            )
                : base(connProv, log, sqlConf, roleRepo)
        {
            _connectionProvider = connProv;
            _log = log;
            _sqlConfiguration = sqlConf;
            _roleRepository = roleRepo;
        }
        // more methods here
    }

This cannot compile because 

    namespace Identity.Dapper.Repositories
    {
        public class UserRepository<TUser, TKey, TUserRole, TRoleClaim> : IUserRepository<TUser, TKey, TUserRole, TRoleClaim>
            where TUser : DapperIdentityUser<TKey>
            where TKey : IEquatable<TKey>
            where TUserRole : DapperIdentityUserRole<TKey>
            where TRoleClaim : DapperIdentityRoleClaim<TKey>

    //....
    }

here we are expecting `TUser : DapperIdentityUser<TKey>` but `TUser : DapperIdentityUser<int, UserClaim, UserRole, UserLogin>`, an ancestor.

Can the repos and stores be altered to use the more general User and Role classes?

**Comment by grandchamp** [comment]
Yes, this is a known issue, i couldn't figure It out How to fix. I always get DI exceptions when i try.

**Comment by nurdyguy** [comment]
I spent a couple of hours working on it yesterday but am stuck on the UserRepository--

    InvalidOperationException: Unable to resolve service for type 
    'SandboxCore.Identity.Dapper.Repositories.Contracts.IRoleRepository`4[Identity.Dapper.Entities.DapperIdentityRole`3[System.Int32,SandboxCore.Identity.Models.UserRole,SandboxCore.Identity.Models.RoleClaim],System.Int32,SandboxCore.Identity.Models.UserRole,SandboxCore.Identity.Models.RoleClaim]' 
    while attempting to activate 
    'SandboxCore.Identity.Dapper.Repositories.UserRepository`6[SandboxCore.Identity.Models.User,System.Int32,SandboxCore.Identity.Models.UserRole,SandboxCore.Identity.Models.RoleClaim,SandboxCore.Identity.Models.UserClaim,SandboxCore.Identity.Models.UserLogin]'.

**Comment by nurdyguy** [comment]
I think I found the solution for this.  It is working on my box anyways...  I'll see if I can put together a pull request.

**Comment by grandchamp** [comment]
Please do it. 
I always received the same exception as you.


Waiting for your PR.

**Comment by nurdyguy** [comment]
Unfortunately you updated for VS2017 which I don't have yet (and can't use because it would conflict with some of my other projects).  As a result I have no intellisense in your project.  Here is my project where I fixed the issue in question:  https://github.com/nurdyguy/SandboxCore  

That project basically copied the repos and stores from your code in and tweaked things until I got it working but it is missing the overloads and also needs some cleanup.  I'll keep working on the PR but it won't be clean.

**Comment by grandchamp** [comment]
Still have the same error on your branch.

**Comment by nurdyguy** [comment]
I hate to say it but I'm not super surprised.  I felt totally blind trying to copy my code in w/o any intellisense or the ability to compile.  My SandboxCore project (linked in my previous comment) works though.  I'll look again and see if I can find what went haywire.

**Comment by grandchamp** [comment]
In parallel i'm working on a branch trying to fix the generics stuff.
Just for clarification, you got the DI exception when trying to use a custom DapperIdentityUser and DapperIdentityRole classes, right?
I'm working on a CustomUser class that inherits from DapperIdentityUser with a Address property inside it (just for customize) and a CustomRole class that adds a IsDummy flag (again, just for customize).
The expected behavior when using the current master version is a DI exception, right?

**Comment by nurdyguy** [comment]
I actually created custom versions of each of the ~/Entities/ classes, though I believe the CustomRole class where the issue lies.  Since the UserRepository requires the RoleRepository to be injected, having a CustomRole rather than the default DapperIdentityRole class causes issues.  In my "fix" I changed the base DapperIdentityRole to be `DapperIdentityRole<TKey, TUserRole, TRoleClaim>` rather than `DapperIdentityRole<TKey>`.  Of course there are a ton of other changes that must go with that, but that was the core of the change.

**Comment by grandchamp** [comment]
Yes, DapperIdentityRole and IRoleRepository are the core.
Please check my commit on https://github.com/grandchamp/Identity.Dapper/commit/1874f1e592a3dfd8e082f500cd56fd1fb60d5150, so far i still having DI exception even by chaging DapperIdentityRole to DapperIdentityRole<TKey, TUserRole, TRoleClaim>.

**Comment by grandchamp** [comment]
I think i solved in https://github.com/grandchamp/Identity.Dapper/commit/0ee1ed47a7110f732a60627b5601568d6434bb24.
What i did was add a `TRole` parameter on `DapperUserStore` and `IUserRepository`, this way the repository knows what type the `Role` is and maps correctly on DI.
The way it was, everytime `IRoleRepository` was being mapped with `DapperIdentityRole<TKey, TUserRole, TRoleClaim>`, so if i changed the `DapperIdentityRole` implementation to a custom one, it won't map correctly, because it was always looking for a unknown register of type `DapperIdentityRole<TKey, TUserRole, TRoleClaim>`.

Can you try and see if it works?

**Comment by nurdyguy** [comment]
I've been working on testing this but no luck so far.  I'll keep you posted.

**Comment by nurdyguy** [comment]
Ok, so since I can't load the project up (whole vs2017 issue) and I've yet to be able to pull the project into a local nuget and load it that way, I settled on pulling a copy of Identity.Dapper and Identity.Dapper.SqlServer directly into my SandboxCore solution, running them in their own projects.  This way I can at least get intellisense and compiling working.  Here is what I've found so far

- Repsoitories/UserRepository.cs has a merge conflict
- Repositories/contracts/IUserRepository.cs has a misspelling line 12:  DapperidentityUserLogin<TKey> (note, this one is my fault, was in my pr)
- Repositories/contracts/IUserRepository.cs lines 17-19, Txxx is not declared
- Stores/DapperUserStore lines 35-37, Txxx is not declared

I basically stopped there because without fixing those things first I can't go much further.  I'm afraid that since I can't do a vs2017 solution making a pull request is pointless.  

In my previous pull request I had to go a bit overboard in declaring my templates in order to deal with the same errors you ran in to.  It wasn't an elegant solution but it seemed to work.  I'll see what I can do on my end to at least deal with the existing errors, even if it means you have to hand merge the code, better than nothing.

**Comment by grandchamp** [comment]
Did you download the generics2 branch?

**Comment by nurdyguy** [comment]
I'm not sure what happened on my last pull down but something in the git repository caused all of those issues.  I pulled fresh and it all went pretty smoothly.  I did have to add a collection extension:

        public static IdentityBuilder AddDapperIdentityForSqlServer<TKey, TUserRole, TRoleClaim, TUserClaim, TUserLogin>(this IdentityBuilder builder)
        {
            builder.Services.AddSingleton<SqlConfiguration, SqlServerConfiguration>();

            AddStores(builder.Services, builder.UserType, builder.RoleType, typeof(TKey), typeof(TUserRole), typeof(TRoleClaim), typeof(TUserClaim), typeof(TUserLogin));

            return builder;
        }

but everything is compiling.  I can't get it to spin up completely yet, I get the following error:

    InvalidOperationException: Unable to resolve service for type 
       'Identity.Dapper.Connections.IConnectionProvider' while attempting to activate      
       'Identity.Dapper.Stores.DapperUserStore`7[SandboxCore.Identity.Models.User,System.Int32,SandboxCore.Identity.Models.UserRole,SandboxCore.Identity.Models.RoleClaim,SandboxCore.Identity.Models.UserClaim,SandboxCore.Identity.Models.UserLogin,SandboxCore.Identity.Models.Role]'.

I'm not quite sure what is up but I'll keep playing with it.

**Comment by grandchamp** [comment]
Did you called `services.ConfigureDapperXXXConnectionProvider()` before `AddIdentity`?

**Comment by nurdyguy** [comment]
Well, I did but then I commented it out at some point and forgot to put it back in...  

I'm spooling now and hitting home page, this looks damn promising.

**Comment by nurdyguy** [comment]
Dude this is way awesome.  A couple of your sql queries needed to be tweaked but that is a separate issue.  the DI issues are gone and that's what this was all about.  I didn't test each of the possible combinations (like using your own User class and UserClaim class but the rest are standard) so there may need to be some tweaking along those lines.  For using all your own classes though this is working perfectly!

**Comment by grandchamp** [comment]
Good to know!
I'll merge my branch to master and publish it on NuGet.
Btw, i'm closing this issue and marking as solved.

Thanks for your help.

**Comment by nurdyguy** [comment]
I was able to extend the default Identity UserManager as well.  After all, what is the point of extending the repo and store if you have no manager methods to access them?  For example, I wanted to be able to find all users who don't have any roles.  You could do that by finding all users, then getting all of the roles for every users etc but truth is a simple left join query would solve this in 1 hop.  In order to extend the UserManager, I created a IUserDataService and UserDataService which extend UserManager.  I ran into some DI issues with the UserStore and repo, not sure why, but solved it with this by altering my Startup.cs to the following:

    services.AddSingleton<IUserStore, UserStore>();
    services.AddSingleton<IRoleStore, RoleStore>();
    services.AddScoped<IUserRepository, UserRepository>();
    services.AddScoped<IRoleRepository, RoleRepository>();

     services.ConfigureDapperSqlServerConnectionProvider(Configuration.GetSection("DapperIdentity"))
            .ConfigureDapperIdentityCryptography(Configuration.GetSection("DapperIdentityCryptography"));

     services.AddIdentity<User, Role>()                        
             .AddDapperIdentityForSqlServer<int, UserRole, RoleClaim, UserClaim, UserLogin>(new SqlServerConfig())
              .AddUserManager<UserDataService>()
               .AddRoleManager<RoleDataService>()
               .AddDefaultTokenProviders();

You can see that I also extended the SqlServerConfiguration.cs with SqlServerConfig.cs.  I did that so I could rename the tables.  I know that new-ing it in there isn't the best way to do that though.  Any suggestions?

**Comment by grandchamp** [comment]
How this works on current Identity implementation?
I've always worked on a way that i use Identity just for login and account related stuff (register, assign roles, etc.), for things not related to this i've always created custom repos/services.

**Comment by nurdyguy** [comment]
I want Identity to keep doing what it does.  I'm just putting a layer over the top to pick up anything Identity isn't handling.  I'll add a couple of repo and service methods on top of Identity's built in to handle extra stuff.  This is just for account related things like user data and roles.  The idea is that I have one userDataService (and a roleDataService) which handle anything and everything I need for user management all with Identity sitting under it, handling the "heavy lifting".  I just add little methods here and there to handle any custom stuff I need (like GetAllUsersWithoutRoles).
